internal/core/worker/operationexecution/operation_execution_worker.go

```diff
@@ -161,65 +167,77 @@ func (w *OperationExecutionWorker) executeOperationFlow(operationID string) erro
+	executionErrCtx := w.workerContext
+	var msg string
 	if strings.Contains(executionErr.Error(), context.Canceled.Error()) {
 		op.Status = operation.StatusCanceled
-
-		loopLogger.Info("operation execution canceled")
-		w.operationManager.AppendOperationEventToExecutionHistory(w.workerContext, op, "Operation canceled by the user")
+		msg = "operation canceled by the user"
+		var cancelFunc func()
+		executionErrCtx, cancelFunc = context.WithTimeout(context.Background(), TerminationGracePeriodTimeout)
+		defer cancelFunc()
 	} else {
 		op.Status = operation.StatusError
-
-		loopLogger.Error("operation execution failed", zap.Error(executionErr))
-		w.operationManager.AppendOperationEventToExecutionHistory(w.workerContext, op, "Operation execution failed")
+		msg = "operation execution failed"
 	}
+	loopLogger.Error(msg, zap.Error(executionErr))
+	w.operationManager.AppendOperationEventToExecutionHistory(executionErrCtx, op, msg)
+	w.rollbackOperation(executionErrCtx, op, def, executionErr, loopLogger, executor)
+	w.finishOperationAndLease(executionErrCtx, op, def, loopLogger)
 }
 
-func (w *OperationExecutionWorker) finishOperationAndLease(op *operation.Operation, def operations.Definition, loopLogger *zap.Logger) {
+func (w *OperationExecutionWorker) finishOperationAndLease(ctx context.Context, op *operation.Operation, def operations.Definition, loopLogger *zap.Logger) {
 	loopLogger.Info("Finishing operation")
 
 	// TODO(gabrielcorado): we need to propagate the error reason.
 	// TODO(gabrielcorado): consider handling the finish operation error.
-	err := w.operationManager.FinishOperation(w.workerContext, op, def)
+	err := w.operationManager.FinishOperation(ctx, op, def)
 	if err != nil {
 		loopLogger.Error("failed to finish operation", zap.Error(err))
 	}
-	err = w.operationManager.RevokeLease(w.workerContext, op)
+	err = w.operationManager.RevokeLease(ctx, op)
 	if err != nil {
 		loopLogger.Error("failed to revoke operation lease", zap.Error(err))
 	}
-	w.operationManager.AppendOperationEventToExecutionHistory(w.workerContext, op, "Operation finished")
+	w.operationManager.AppendOperationEventToExecutionHistory(ctx, op, "Operation finished")
 }
 
-func (w OperationExecutionWorker) executeOperationWithLease(operationContext context.Context, op *operation.Operation, def operations.Definition, executor operations.Executor) error {
-	return w.executeCollectingLatencyMetrics(op.DefinitionName, func() error {
-		return executor.Execute(operationContext, op, def)
-	})
+func (w OperationExecutionWorker) executeOperationWithLease(operationContext context.Context, op *operation.Operation, def operations.Definition, executor operations.Executor) (err error) {
+	done := make(chan error, 1)
+	go func() {
+		done <- w.executeCollectingLatencyMetrics(op.DefinitionName, func() error {
+			return executor.Execute(operationContext, op, def)
+		})
+	}()
+	select {
+	case <-operationContext.Done():
```

Comment:
i would vote for handling context cancelation in the function that actually calls this method

why? because its the caller responsibility to deal with that i would say